design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+- `environment`: contains a list of `EnvironmentConfigs` that can be referenced
+  by specific types of patches and a list of`EnvironmentPatches` . Also an
+  `alpha` feature, similar implications to what already said for `functions`,
+  would require additional information and can invalidate the assumption that
+  we render valid Composed Resources if defined.
+- `writeConnectionSecretsToNamespace` : the name of the namespace the `Secret`
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. The CRDs of the Managed Resources defined as `base` and traversed by
+   `patches` might not yet have been created when the Composition is being
+   validated.
+2. The CRD of the Composite Resources specified by `CompositeTypeRef` and
+   traversed by `patches` might not yet have been created when the Composition
+   is being validated.
+3. We do not have all the required CRDs locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+In order to address the possible absence of Managed Resources' CRDs, Composite
+Resources' CRDs or any other required external resource at validation time we
+should allow explicitly setting one of 2 modes, `strict` and `loose`(default) .
+These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that any missing resource
+should result in a direct rejection or a possibly incomplete validation.
+
+In `loose` mode we should emit warnings in case any of the preconditions for a
+check are not met, and skip further validation related to that check.
+
+Regarding the approach, I propose to implement a Composition Validating Webhook
+by validating each `resource` and final rendered `Composed
+Resource`, querying the API Server to get the needed schemas.
```

Comment:
Worth noting that in future I could imagine a composition function validating the final XR before apply, as long as it was the last function in the chain.

Reply:
Makes sense, that's why Functions should be considered non deterministic and more complex validation logic won't make sense in their presence as suggested below: https://github.com/crossplane/crossplane/pull/3756/files#diff-8dfb220f48da6208755c36731275d4d8ff51209399fedbffd3c2ea3e073fa55fR204